pkg/consensus/retry_watcher.go

```diff
@@ -30,16 +34,29 @@ type retryWatcher struct {
 	getOpts   *etcd.GetOptions
 	watchOpts *etcd.WatcherOptions
 
+	refreshPeriod time.Duration
+	refreshNext   time.Time
+	now           func() time.Time
+
 	cur etcd.Watcher
 }
 
 func (w *retryWatcher) Next(ctx context.Context) (*etcd.Response, error) {
+	// Periodically force a full tree refresh.
+	if w.now().After(w.refreshNext) &&
+		!w.refreshNext.IsZero() && w.refreshPeriod != 0 {
+		w.cur = nil
+	}
+
 	if w.cur != nil {
 		r, err := w.cur.Next(ctx)
 
 		if etcdErr, ok := err.(etcd.Error); ok {
 			// If the error code indicates that further Next() attempts will fail,
 			// clear the current etcd.Watcher to force a full tree refresh.
+			// Ignore all other errors. For silent errors, upper-bound the
+			// effects with |refreshPeriod| by defensively forcing
+			// a full tree refresh periodically.
 			if etcdErr.Code == etcd.ErrorCodeEventIndexCleared ||
 				etcdErr.Code == etcd.ErrorCodeWatcherCleared {
 				w.cur = nil
```

Comment:
Not your code, but the recursive call to `w.Next` below doesn't seem like a great idea to me, vis a vis just returning the error like any other. That removes the possibility of silent errors; `Allocator` will log a warning and retry `Next` itself.

Reply:
(FWIW, I don't think there actually are any other silent errors; the other Etcd errors are driven by bad user input and we can expect to see a stack overflow, but I could be wrong here).